22.2.3+ (#1)
friedcucumber/maui-authenticate: README.md

# Authenticate Users with the DevExpress Web API Service

> This example shows how to use our free [.NET App Security Library & Web API Service](https://www.devexpress.com/products/net/application_framework/security-web-api-service.xml) to implement authentication in your .NET MAUI application. A wizard help you generate a ready-to-use authentication service. This service uses the [Entity Framework Core](https://docs.microsoft.com/en-us/ef/core/) ORM to access a database.    

Reviewer: ```suggestion
This example shows how to use our free [.NET App Security Library & Web API Service](https://www.devexpress.com/products/net/application_framework/security-web-api-service.xml) to implement authentication in your .NET MAUI application. A wizard helps you generate a ready-to-use authentication service. This service uses the [Entity Framework Core](https://docs.microsoft.com/en-us/ef/core/) ORM to access a database.    
```

        <!-* ... -->
    </sites>
    ```

3. Right-click the MAUI project, choose `Set as Startup Project`, and select your emulator. Note that physical devices that are attached over USB do not allow you to access your machine's localhost.
4. Right-click the `WebAPI` project and select `Debug > Start new instance`.
5. Right-click the `MAUI` project and select `Debug > Start new instance`.

## Implementation Details

### Service and Communication

* DevExpress Web API Service uses JSON Web Tokens (JWT) to authorize users. Call `WebAPI`'s **Authenticate** endpoint and pass a username and password to the endpoint from the .NET MAUI application. In this example, token generation logic is implemented in the `WebAPIService.RequestTokenAsync` method:
  
    ```csharp
      private async Task<HttpResponseMessage> RequestTokenAsync(string userName, string password) {
            return await HttpClient.PostAsync($"{ApiUrl}Authentication/Authenticate",
                                                new StringContent(JsonSerializer.Serialize(new { userName, password = $"{password}" }), Encoding.UTF8,
                                                ApplicationJson));
      }
    ```

>     Include the token into [HttpClient.DefaultRequestHeaders.Authorization](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httprequestheaders.authorization?view=net-7.0). All subsequent requests will be able to access private endpoints and data: 

Reviewer: ```suggestion
    Include the token in [HttpClient.DefaultRequestHeaders.Authorization](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httprequestheaders.authorization?view=net-7.0). All subsequent requests can then access private endpoints and data: 
```

    File to Look At: [LoginPage.xaml](CS/MAUI/Views/LoginPage.xaml)
    
    ```csharp
    public class LoginViewModel : BaseViewModel {
        // ...
        public LoginViewModel() {
            LoginCommand = new Command(OnLoginClicked);
            SignUpCommand = new Command(OnSignUpClicked);
            PropertyChanged +=
                (_, __) => LoginCommand.ChangeCanExecute();

        }
        // ...
        public Command LoginCommand { get; }
        public Command SignUpCommand { get; }
        // ...
    }
    ```

    File to Look At: [LoginViewModel.cs](CS/MAUI/ViewModels/LoginViewModel.cs)

### Debug Specifics

Android emulator and iOS simulator request a certificate to access a service over HTTPS. In this example, we switch to HTTP in debug mode:

```csharp
#if !DEBUG
    app.UseHttpsRedirection();
#endif
```

#### MAUI - Android

```xml
<network-security-config>
    <domain-config cleartextTrafficPermitted="true">
        <domain includeSubdomains="true">10.0.2.2</domain>
    </domain-config>
</network-security-config>
```

#### MAUI - iOS

```xml
<key>NSAppTransportSecurity</key>
<dict>
    <key>NSAllowsLocalNetworking</key>
    <true/>
</dict>
```

> This allows you to bypass the certificate check. You don't need to create a development certificate or implement HttpClient handlers.

Reviewer: ```suggestion
This allows you to bypass the certificate check without the need to create a development certificate or implement HttpClient handlers.
```

For more information, please refer to [Connect to local web services from Android emulators and iOS simulators](https://learn.microsoft.com/en-us/dotnet/maui/data-cloud/local-web-services?view=net-maui-7.0#android-network-security-configuration).

> **We highly recommend that you use HTTP only when you develop/debug your application. In production, use HTTPS for security reasons.**

Reviewer: ```suggestion
**We recommend that you use HTTP only when you develop/debug your application. In production, use HTTPS for security reasons.**
```

> # Authenticate Users with the WebAPI Service

Reviewer 2: ```suggestion
# Authenticate Users with the DevExpress Web API Service
```

> This example shows how to use our free [.NET App Security Library](https://www.devexpress.com/products/net/application_framework/security-web-api-service.xml) to implement authentication in your .NET MAUI application. The [.NET App Security Library](https://www.devexpress.com/products/net/application_framework/security-web-api-service.xml) includes a wizard that generates a ready-to-use authentication service. This service uses the [Entity Framework Core](https://docs.microsoft.com/en-us/ef/core/) ORM to access a database.    

Reviewer 2: ```suggestion
This example shows how to use our free [.NET App Security Library & Web API Service](https://www.devexpress.com/products/net/application_framework/security-web-api-service.xml) to implement authentication in your .NET MAUI application. A wizard help you generate a ready-to-use authentication service. This service uses the [Entity Framework Core](https://docs.microsoft.com/en-us/ef/core/) ORM to access a database.    
```

> You can find more information about our WebAPI Service's access restrictions in the following topic and video:

Reviewer 2: ```suggestion
You can find more information about our Web API Service's access restrictions in the following resources:
```

> [Create a Standalone Web API Application](https://docs.devexpress.com/eXpressAppFramework/403401/backend-web-api-service/create-new-application-with-web-api-service?p=net6)

Reviewer 2: ```suggestion
[Article: Create a Standalone Web API Application](https://docs.devexpress.com/eXpressAppFramework/403401/backend-web-api-service/create-new-application-with-web-api-service?p=net6)
```

> [A 1-Click Solution for CRUD Web API with Role-based Access Control via EF Core & ASP.NET](https://www.youtube.com/watch?v=T7y4gwc1n4w&list=PL8h4jt35t1wiM1IOux04-8DiofuMEB33G)

Reviewer 2: ```suggestion
[Video: A 1-Click Solution for CRUD Web API with Role-based Access Control via EF Core & ASP.NET](https://www.youtube.com/watch?v=T7y4gwc1n4w&list=PL8h4jt35t1wiM1IOux04-8DiofuMEB33G)
```

## Prerequisites

[SQL Server](https://www.microsoft.com/en-us/sql-server/sql-server-downloads), if you run this solution on Windows.

## Run Projects

> Perform the following steps to run the project:

Reviewer 2: We don't need this sentence, if it doesn't add anything to the title above it. 

```suggestion

```

> 1. Open the `WebAPI` solution in the Visual Studio as an administrator. This is required to create the database whose settings are defined in the `appsettings.json` file.

Reviewer 2: ```suggestion
1. Open the `WebAPI` solution. Run Visual Studio as an administrator so that the IDE can create the database as defined in `appsettings.json`.
```

> 2. Choose the `WebApi` item in the **debug** dropdown menu. From this menu, you can debug the project on the [Kestrel](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-7.0) web server.

Reviewer 2: ```suggestion
2. Choose the `WebApi` item in the **debug** dropdown menu. This selection will debug the project on the [Kestrel](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-7.0) web server.
```

    ![Run Settings](images/authenticate-run-settings@2x.png)

>     If you prefer **IIS Express** instead of `Kestrel`, you need to select the IIS Express on the **debug** dropdown menu, open the `.vs\MAUI_WebAPI\config\applicationhost.config` in an external text editor, and add the following code:

Reviewer 2: I'm moving a long file name to the end of the sentence:

```suggestion
    If you prefer IIS Express to Kestrel, select **IIS Express** on the **debug** dropdown menu. Use an external text editor to add the following code to `.vs\MAUI_WebAPI\config\applicationhost.config`:
```

    ```xaml
    <sites>
        <site name="WebSite1" id="1" serverAutoStart="true">
        <!-* ... -->
            <bindings>
                <binding protocol="http" bindingInformation="*:65201:*" />
                <binding protocol="https" bindingInformation="*:44317:*" />
                <binding protocol="https" bindingInformation="*:44317:localhost" />
                <binding protocol="http" bindingInformation="*:65201:localhost" />
            </bindings>
        </site>
        <!-* ... -->
    </sites>
    ```

> 3. Right-click the MAUI project, choose `Set as Startup Project` and select your emulator. Note that physical devices that are attached over USB do not allow you to access your machine's localhost.

Reviewer 2: ```suggestion
3. Right-click the MAUI project, choose `Set as Startup Project`, and select your emulator. Note that physical devices that are attached over USB do not allow you to access your machine's localhost.
```

> 4. Right-click the `WebAPI` project and select `Debug > Start new instance` to run the `WebAPI` project.

Reviewer 2: ```suggestion
4. Right-click the `WebAPI` project and select `Debug > Start new instance`.
```

5. Right-click the `MAUI` project and select `Debug > Start new instance` to run the `MAUI` project.

## Implementation Details

### Service and Communication

* The WebAPI service uses json web tokens (JWT) to authorize users. Call `WebAPI`'s **Authenticate** endpoint and pass a username and password to the endpoint from the .NET MAUI application. In the example, this logic is implemented in the `WebAPIService.RequestTokenAsync` method:
  
    ```csharp
      private async Task<HttpResponseMessage> RequestTokenAsync(string userName, string password) {
            return await HttpClient.PostAsync($"{ApiUrl}Authentication/Authenticate",
                                                new StringContent(JsonSerializer.Serialize(new { userName, password = $"{password}" }), Encoding.UTF8,
                                                ApplicationJson));
      }
    ```

>     Include the token in request’s [HttpClient.DefaultRequestHeaders.Authorization](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httprequestheaders.authorization?view=net-7.0) header to allow all subsequent requests access private endoints and data: 

Reviewer 2: ```suggestion
    Include the token into [HttpClient.DefaultRequestHeaders.Authorization](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httprequestheaders.authorization?view=net-7.0). All subsequent requests will be able to access private endpoints and data: 
```

    File to Look At: [LoginPage.xaml](CS/MAUI/Views/LoginPage.xaml)
    
    ```csharp
    public class LoginViewModel : BaseViewModel {
        // ...
        public LoginViewModel() {
            LoginCommand = new Command(OnLoginClicked);
            SignUpCommand = new Command(OnSignUpClicked);
            PropertyChanged +=
                (_, __) => LoginCommand.ChangeCanExecute();

        }
        // ...
        public Command LoginCommand { get; }
        public Command SignUpCommand { get; }
        // ...
    }
    ```

    File to Look At: [LoginViewModel.cs](CS/MAUI/ViewModels/LoginViewModel.cs)

### Debug Specifics

Android emulator and iOS simulator request a certificate to access a service over HTTPS. In this example, we switch to HTTP in debug mode:

```csharp
#if !DEBUG
    app.UseHttpsRedirection();
#endif
```

#### MAUI - Android

```xml
<network-security-config>
    <domain-config cleartextTrafficPermitted="true">
        <domain includeSubdomains="true">10.0.2.2</domain>
    </domain-config>
</network-security-config>
```

#### MAUI - iOS

```xml
<key>NSAppTransportSecurity</key>
<dict>
    <key>NSAllowsLocalNetworking</key>
    <true/>
</dict>
```

> This allows you to bypass the certificate check and avoid creating a development certificate and implementing HttpClient handlers.

Reviewer 2: ```suggestion
This allows you to bypass the certificate check. You don't need to create a development certificate or implement HttpClient handlers.
```

For more information, please refer to [Connect to local web services from Android emulators and iOS simulators](https://learn.microsoft.com/en-us/dotnet/maui/data-cloud/local-web-services?view=net-maui-7.0#android-network-security-configuration).

> **We highly recommend using HTTP only for debug and development purposes. In production, use HTTPS for security reasons.**

Reviewer 2: ```suggestion
**We highly recommend that you use HTTP only when you develop/debug your application. In production, use HTTPS for security reasons.**
```
